core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -422,11 +425,17 @@ private[log] class Cleaner(val id: Int,
       for (entry <- messages.shallowIterator) {
         val size = MessageSet.entrySize(entry.message)
         stats.readMessage(size)
-        if (entry.message.compressionCodec == NoCompressionCodec) {
+        val sourceCodec: CompressionCodec = entry.message.compressionCodec
+        val targetCodec = BrokerCompressionCodec.getTargetCompressionCodec(brokerCompressionType, sourceCodec)
+        if (sourceCodec == NoCompressionCodec) {
           if (shouldRetainMessage(source, map, retainDeletes, entry)) {
             val convertedMessage = entry.message.toFormatVersion(messageFormatVersion)
-            ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
-            stats.recopyMessage(size)
+            if (targetCodec == NoCompressionCodec) { // same as sourceCodec
+              ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
+              stats.recopyMessage(size)
+            } else {
+              compressMessages(writeBuffer, targetCodec, messageFormatVersion, Seq(new MessageAndOffset(convertedMessage, entry.offset)))
```

Comment:
It seems consumer is no longer a concern any more for FetchRequest v3 and above because of KIP-74. i.e. the consumers will be able to fetch the messages even if they are bigger than max message size after recompression. 

Theoretically it is possible for a single message smaller than max message size with one compression type to grow beyond the max message size after changing to another compression type. There seems not much we can do if that happened.